rust-runtime/aws-smithy-runtime/src/client/retries/classifiers.rs

```diff
@@ -119,45 +145,86 @@ impl HttpStatusCodeClassifier {
             retryable_status_codes: retryable_status_codes.into(),
         }
     }
+
+    /// Return the priority of this retry classifier.
+    pub fn priority() -> RetryClassifierPriority {
+        RetryClassifierPriority::http_status_code_classifier()
+    }
 }
 
 impl ClassifyRetry for HttpStatusCodeClassifier {
-    fn classify_retry(&self, ctx: &InterceptorContext) -> Option<RetryReason> {
-        ctx.response()
+    fn classify_retry(&self, ctx: &InterceptorContext) -> RetryAction {
+        let is_retryable = ctx
+            .response()
             .map(|res| res.status().as_u16())
             .map(|status| self.retryable_status_codes.contains(&status))
-            .unwrap_or_default()
-            .then_some(RetryReason::Error(ErrorKind::TransientError))
+            .unwrap_or_default();
+
+        if is_retryable {
+            RetryAction::Retry(ErrorKind::TransientError)
+        } else {
+            RetryAction::DontCare
+        }
     }
 
     fn name(&self) -> &'static str {
         "HTTP Status Code"
     }
+
+    fn priority(&self) -> RetryClassifierPriority {
+        Self::priority()
+    }
 }
 
-// Generic smithy clients would have something like this:
-// pub fn default_retry_classifiers() -> RetryClassifiers {
-//     RetryClassifiers::new()
-//         .with_classifier(SmithyErrorClassifier::new())
-//         .with_classifier(ModeledAsRetryableClassifier::new())
-//         .with_classifier(HttpStatusCodeClassifier::new())
-// }
-// This ordering is different than the default AWS ordering because the old generic client classifier
-// was the same.
+/// Given an iterator of retry classifiers and an interceptor context, run retry classifiers on the
+/// context. Each classifier is passed the classification result from the previous classifier (the
+/// 'root' classifier is passed `None`.)
+pub fn run_classifiers_on_ctx(
+    classifiers: impl Iterator<Item = SharedRetryClassifier>,
+    ctx: &InterceptorContext,
+) -> RetryAction {
+    // By default, don't retry
+    let mut result = RetryAction::DontCare;
+
+    for classifier in classifiers {
+        let new_result = classifier.classify_retry(ctx);
+
+        // If the result is `DontCare`, continue to the next classifier
+        // without overriding any previously-set result.
+        if new_result == RetryAction::DontCare {
+            continue;
+        }
+
+        // Otherwise, set the result to the new result.
+        tracing::trace!(
+            "Classifier '{}' set the result of classification to '{}'",
+            classifier.name(),
+            new_result
+        );
+        result = new_result;
+
+        // If the result is `RetryForbidden`, stop running classifiers.
+        if result == RetryAction::RetryForbidden {
+            break;
```

Comment:
I think we should probably log something here

Reply:
I'll add a trace log